<a href="https://colab.research.google.com/github/solver-Mart1n/data-science/blob/solver-Mart1n-c06w1p2-basic-sql/3_Lab_INSERT_UPDATE_DELETE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1> COUNT, DISTINCT, LIMIT in SQL using Jupyter Notebooks </h1>

## Description
In this lab, you will learn some commonly used DML (Data Manipulation Language) statements of SQL other than SELECT. First, you will learn the
INSERT statement, which is used to insert new rows into a table. Next, you will learn the UPDATE statement which is used to update the data in
existing rows in the table. Lastly, you will learn the DELETE statement which is used to remove rows from a table.

<h3>Objectives</h3>
<h4>After completing this lab, you will be able to:</h4>
    <ul>
        <li>Insert new rows into a table</li>
        <li>Update data in existing rows of the table</li>
        <li>Remove rows from a table</li>
    </ul>


<h3>Table of Contents</h3>
    <ul>
        <li>Building the Database from an Internet Source</li>
        <li>Exploring the Database</li>
        <li>Using INSERT Statement</li>
        <li>Using UPDATE Statement</li>
        <li>Using DELETE Statement</li>
        <li>Practice Exercises INSERT, UPDATE, DELETE</li>
    </ul>

<p>Estimated Time Needed: <strong>30 min</strong></p>
<hr>

## Building the Database from an Internet Source
The database used in this lab comes from the following dataset source: [Film Locations in San Francisco](https://data.sfgov.org/Culture-and-Recreation/Film-Locations-in-San-Francisco/yitu-d5am/about_data) under a [PDDL: Public Domain Dedication and License](http://opendatacommons.org/licenses/pddl/1.0/).

### Ingesting a CSV from a Data Source Endpoint
Three API parameters are used on top of the base URL for the data source. This data paging is in compliance with the row limit of the endpoint API.

In [1]:
url1 = 'https://data.sfgov.org/resource/yitu-d5am.csv?$limit=1000&$offset=0'
url2 = 'https://data.sfgov.org/resource/yitu-d5am.csv?$limit=1000&$offset=999'
url3 = 'https://data.sfgov.org/resource/yitu-d5am.csv?$limit=49&$offset=1999'

The three will be ingested as data frames and combined into one. And this will be used to generate a CSV.

In [2]:
!pip install pandas==1.3.3

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.5/11.5 MB 28.7 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.0.3
    Uninstalling pandas-2.0.3:
      Successfully uninstalled pandas-2.0.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
bigframes 1.4.0 requires pandas>=1.5.0, but you have pandas 1.3.3 which is incompatible.
google-colab 1.0.0 requires pandas==2.0.3, but you have pandas 1.3.3 which is incompatible.
mizani 0.9.3 requires pandas>=1.3.5, but you have pandas 1.3.3 which is incompatible.
plotnine 0.12.4 requires pandas>=1.5.0, but you have pandas 1.3.3 which is incompatible.
statsmodels 0.14.2 requires pandas!=2.1.0,>=1.4, but you have pandas 1.3.3 which is incompatible.
xarray 2023.7.0 requires pandas>=1.4, but you have pandas 1.3.3 which is incompatible.


In [3]:
import pandas as pd

In [4]:
# Read the dataset from a csv file
df1 = pd.read_csv(url1, header=0, sep=",")

# Display the first few rows of the DataFrame
df1.tail()

title  release_year  \
995            The Dead Pool          1988   
996           Vegas in Space          1992   
997  Looking Season 2 ep 208          2015   
998        Tales of the City          2019   
999            The Dead Pool          1988   

                                             locations fun_facts  \
995      Pierce & California Streets (Pacific Heights)       NaN   
996  Under Highway 101 (near Potrero and Cesar Chav...       NaN   
997                                 1246 Folsom Street       NaN   
998                City Lights Books, 261 Columbus Ave       NaN   
999          McAllister & Scott Streets (Hayes Valley)       NaN   

                  production_company               distributor  \
995            Warner Bros. Pictures     Warner Bros. Pictures   
996              Troma Entertainment  MCA/Universal Home Video   
997  Mission Street Productions, LLC                       HBO   
998         Universal Television LLC                   Netflix   
999            Warner Bros. Pictures     Warner Bros. Pictures   

                                   director  \
995                          Buddy Van Horn   
996                         Phillip R. Ford   
997                            Andrew Haigh   
998  Alan Poul, Silas Howard, Stacie Passon   
999                          Buddy Van Horn   

                                                writer         actor_1  \
995                                  Harry Julian Fink  Clint Eastwood   
996                                         Doris Fish      Doris Fish   
997                                     Michael Lannan  Jonathan Groff   
998  Armistead Maupin, Lauren Morelli, Hansol Jung,...    Laura Linney   
999                                  Harry Julian Fink  Clint Eastwood   

             actor_2            actor_3  :@computed_region_6qbp_sg9q  \
995      Liam Neeson  Patricia Clarkson                        102.0   
996           Miss X       Ginger Quest                         60.0   
997  Frankie Alvarez    Murray Bartlett                         32.0   
998       Ellen Page    Murray Bartlett                        104.0   
999      Liam Neeson  Patricia Clarkson                         97.0   

     :@computed_region_ajp5_b2md  :@computed_region_26cr_cadq  
995                         30.0                          6.0  
996                          2.0                          9.0  
997                         34.0                         10.0  
998                          6.0                          3.0  
999                         39.0                         11.0

In [5]:
# Read the dataset from a csv file
df2 = pd.read_csv(url2, header=0, sep=",")

# Display the first few rows of the DataFrame
df2.head()

title  release_year  \
0                   The Dead Pool          1988   
1                           Venom          2018   
2  A Taiwanese Tale of Two Cities          2018   
3        The Matrix Resurrections          2021   
4         Looking Season 2 ep 202          2015   

                                   locations fun_facts  \
0  McAllister & Scott Streets (Hayes Valley)       NaN   
1                     Hyde St at Chestnut St       NaN   
2                     Sutter St at Taylor St       NaN   
3                                 1 Front St       NaN   
4                            314 11th Street       NaN   

                production_company  \
0            Warner Bros. Pictures   
1         L.O.Z. Productions, Inc.   
2        Envision Productions Inc.   
3             Adobe Pictures, Inc.   
4  Mission Street Productions, LLC   

                                  distributor         director  \
0                       Warner Bros. Pictures   Buddy Van Horn   
1  Columbia Pictures, Sony Pictures Releasing  Ruben Fleischer   
2                                     Netflix     Tien-Lun Yeh   
3                             Warner Brothers   Lana Wachowski   
4                                         HBO     Andrew Haigh   

                                              writer         actor_1  \
0                                  Harry Julian Fink  Clint Eastwood   
1                      Jeff Pinkner, Scott Rosenberg       Tom Hardy   
2  Ling-Hui Chen, Nancy Chen, Chih-Chi Fan, Chia-...     Tammy Cheng   
3   Lana Wachowski, David Mitchell, Aleksander Hemon    Keanu Reeves   
4                                     Michael Lannan  Jonathan Groff   

            actor_2              actor_3  :@computed_region_6qbp_sg9q  \
0       Liam Neeson    Patricia Clarkson                         97.0   
1  Michelle Wiliams            Riz Ahmed                        107.0   
2       Peggy Tseng         Shen-Hao Wen                         50.0   
3      Keanu Reeves  Neil Patrick Harris                        108.0   
4   Frankie Alvarez      Murray Bartlett                         32.0   

   :@computed_region_ajp5_b2md  :@computed_region_26cr_cadq  
0                         39.0                         11.0  
1                         32.0                          6.0  
2                         21.0                          3.0  
3                          8.0                          3.0  
4                         20.0                         10.0

In [6]:
df2.tail()

title  release_year  \
995             The Caine Mutiny          1954   
996                     GirlBoss          2017   
997                      Looking          2014   
998  The House on Telegraph Hill          1951   
999              Women is Losers          2020   

                               locations  \
995                      The Embarcadero   
996  Boulevard Restaurant, 1 Mission St.   
997                        2866 24th St.   
998                           Coit Tower   
999                         1080 Capp St   

                                             fun_facts  \
995                                                NaN   
996                                                NaN   
997                                                NaN   
998  The Tower was funded by a gift bequeathed by L...   
999                                                NaN   

                   production_company                       distributor  \
995        Stanley Kramer Productions                 Columbia Pictures   
996        Hippolyta Productions, LLC                           Netflix   
997   Mission Street Productions, LLC             Home Box Office (HBO)   
998  Twentieth Century Fox Film Corp.  Twentieth Century Fox Film Corp.   
999         Look at the Moon Pictures                               NaN   

                                              director              writer  \
995                                     Edward Dmytryk     Stanley Roberts   
996  Jamie Babbit, Amanda Brotchie, Steven K. Tsuch...          Kay Cannon   
997                                       Andrew Haigh      Michael Lannan   
998                                        Robert Wise          Elick Moll   
999                                 Lissette Feliciano  Lissette Feliciano   

              actor_1             actor_2           actor_3  \
995   Humphrey Bogart      Fred MacMurray       Jose Ferrer   
996   Britt Robertson          Ellie Reed        Amanda Rea   
997    Jonathan Groff  Frankie J. Alvarez   Murray Bartlett   
998  Richard Basehart   Valentina Cortese  William Lundigan   
999      Lorenza Izzo            Simu Liu         Liza Weil   

     :@computed_region_6qbp_sg9q  :@computed_region_ajp5_b2md  \
995                         99.0                         23.0   
996                        108.0                          8.0   
997                         53.0                         20.0   
998                         18.0                         23.0   
999                         53.0                         20.0   

     :@computed_region_26cr_cadq  
995                          3.0  
996                         10.0  
997                          2.0  
998                          3.0  
999                          2.0

In [7]:
# Read the dataset from a csv file
df3 = pd.read_csv(url3, header=0, sep=",")

# Display the first few rows of the DataFrame
df3.head()

title  release_year  \
0     Looking "Special"          2016   
1               Looking          2014   
2           Dirty Harry          1971   
3  Terminator - Genisys          2015   
4             Foul Play          1978   

                                           locations  \
0                                 1080 Natoma Street   
1                                        2000 Folsom   
2                   Mt. Davidson Cross, Mt. Davidson   
3                                    Market at Mason   
4  Potrero Hill Bathhouse & Feed Co. (199 Mississ...   

                                           fun_facts  \
0                                                NaN   
1                                                NaN   
2  The cross sits at the highest point in San Fra...   
3                                                NaN   
4                                                NaN   

                production_company            distributor       director  \
0  Mission Street Productions, LLC                    HBO   Andrew Haigh   
1  Mission Street Productions, LLC  Home Box Office (HBO)   Andrew Haigh   
2              The Malpaso Company        Warner Brothers     Don Siegel   
3               T5 Productions LLC     Paramount Pictures    Alan Taylor   
4               Paramount Pictures     Paramount Pictures  Colin Higgins   

              writer                actor_1             actor_2  \
0     Michael Lannan         Jonathan Groff     Frankie Alvarez   
1     Michael Lannan         Jonathan Groff  Frankie J. Alvarez   
2  Harry Julian Fink         Clint Eastwood      Harry Guardino   
3      James Cameron  Arnold Schwarzenegger        Jason Clarke   
4      Colin Higgins            Goldie Hawn         Chevy Chase   

            actor_3  :@computed_region_6qbp_sg9q  :@computed_region_ajp5_b2md  \
0   Murray Bartlett                         32.0                         20.0   
1   Murray Bartlett                         53.0                         20.0   
2               NaN                        117.0                         41.0   
3               NaN                         20.0                         34.0   
4  Burgess Meredith                         54.0                         26.0   

   :@computed_region_26cr_cadq  
0                         10.0  
1                          2.0  
2                          8.0  
3                         10.0  
4                          9.0

In [8]:
df3.tail()

title  release_year  \
44  Murder in the First, Season 2          2015   
45                    I’m A Virgo          2023   
46                       Superman          1978   
47  Murder in the First, Season 3          2016   
48           George of the Jungle          1997   

                                            locations  \
44  Harriet Street between Bryant and Brannan Streets   
45       Columbus Avenue between Pacifiic and Jackson   
46                                 Golden Gate Bridge   
47                                2633 Vallejo Street   
48  Neiman Marcus (150 Stockton Street, Union Square)   

                                            fun_facts  \
44                                                NaN   
45                                                NaN   
46  With 23 miles of ladders and 300,000 rivets in...   
47                                                NaN   
48  The City of Paris department store previously ...   

                 production_company                      distributor  \
44  Turner North Center Productions  Turner Network Television (TNT)   
45        Bing Bang Productions Inc                      Prime Video   
46                   Dovemead Films            Warner Bros. Pictures   
47  Turner North Center Productions  Turner Network Television (TNT)   
48             Walt Disney Pictures             Buena Vista Pictures   

          director        writer            actor_1             actor_2  \
44  Steven Bochcho    Eric Lodal         Taye Diggs  Kathleen Robertson   
45     Boots Riley   Boots Riley     Jharrel Jerome           Mike Epps   
46  Richard Donner  Jerry Siegel  Christopher Reeve        Gene Hackman   
47  Steven Bochcho    Eric Lodal         Taye Diggs  Kathleen Robertson   
48     Sam Weisman      Jay Ward     Brendan Fraser         Leslie Mann   

             actor_3  :@computed_region_6qbp_sg9q  \
44  Ian Anthony Dale                         32.0   
45      Carmen Ejogo                         77.0   
46     Marlon Brando                          NaN   
47  Ian Anthony Dale                        102.0   
48               NaN                         19.0   

    :@computed_region_ajp5_b2md  :@computed_region_26cr_cadq  
44                         34.0                         10.0  
45                          8.0                          3.0  
46                          NaN                          NaN  
47                         30.0                          6.0  
48                          8.0                          3.0

Visible from the tail() and head() print outs of the three data frames, indices start from 0 each time a CSV is loaded. The "ignore_index" attribute of the data frame append() function is set to _True_ in order to avoid copying repeating indices to the aggregated data frame.

In [9]:
df = df1.append(df2,ignore_index=True)
df = df.append(df3,ignore_index=True)
len(df)

2049

In [10]:
df.columns

Index(['title', 'release_year', 'locations', 'fun_facts', 'production_company',
       'distributor', 'director', 'writer', 'actor_1', 'actor_2', 'actor_3',
       ':@computed_region_6qbp_sg9q', ':@computed_region_ajp5_b2md',
       ':@computed_region_26cr_cadq'],
      dtype='object')

In [11]:
df.drop(columns=[':@computed_region_6qbp_sg9q', ':@computed_region_ajp5_b2md', ':@computed_region_26cr_cadq'],inplace=True)

In [12]:
df.set_axis(['Title', 'ReleaseYear', 'Locations', 'FunFacts', 'ProductionCompany', 'Distributor', 'Director', 'Writer', 'Actor1', 'Actor2', 'Actor3'], axis='columns', inplace=True)

As a result, the index labels of the combined data frame starts from 0 and ends with 2048. This indicates a successful merge of the pages of data from the [Data SF](https://data.sfgov.org/resource/yitu-d5am.csv) source URL.

In [13]:
df

Title  ReleaseYear  \
0              Experiment in Terror         1962   
1              Experiment in Terror         1962   
2                   Chan is Missing         1982   
3                  A View to a Kill         1985   
4                  The Californians         2005   
...                             ...          ...   
2044  Murder in the First, Season 2         2015   
2045                    I’m A Virgo         2023   
2046                       Superman         1978   
2047  Murder in the First, Season 3         2016   
2048           George of the Jungle         1997   

                                              Locations  \
0           The Sea Captain's Chest (Fisherman's Wharf)   
1                                100 St. Germain Avenue   
2     Li Po (916 Grant Avenue at Washington, Chinatown)   
3      Taylor and Jefferson Streets (Fisherman's Wharf)   
4                                                   NaN   
...                                                 ...   
2044  Harriet Street between Bryant and Brannan Streets   
2045       Columbus Avenue between Pacifiic and Jackson   
2046                                 Golden Gate Bridge   
2047                                2633 Vallejo Street   
2048  Neiman Marcus (150 Stockton Street, Union Square)   

                                               FunFacts  \
0                                                   NaN   
1                                                   NaN   
2                                                   NaN   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
2044                                                NaN   
2045                                                NaN   
2046  With 23 miles of ladders and 300,000 rivets in...   
2047                                                NaN   
2048  The City of Paris department store previously ...   

                    ProductionCompany                      Distributor  \
0       Columbia Pictures Corporation                Columbia Pictures   
1       Columbia Pictures Corporation                Columbia Pictures   
2                    New Yorker Films                 New Yorker Films   
3                 Metro-Goldwyn Mayer     MGM/UA Entertainment Company   
4                 Parker Film Company                Fabrication Films   
...                               ...                              ...   
2044  Turner North Center Productions  Turner Network Television (TNT)   
2045        Bing Bang Productions Inc                      Prime Video   
2046                   Dovemead Films            Warner Bros. Pictures   
2047  Turner North Center Productions  Turner Network Television (TNT)   
2048             Walt Disney Pictures             Buena Vista Pictures   

             Director                                Writer  \
0       Blake Edwards                           The Gordons   
1       Blake Edwards                           The Gordons   
2          Wayne Wang                            Wayne Wang   
3           John Glen                       Richard Maibaum   
4     Jonathan Parker  Jonathan Parker & Catherine DiNapoli   
...               ...                                   ...   
2044   Steven Bochcho                            Eric Lodal   
2045      Boots Riley                           Boots Riley   
2046   Richard Donner                          Jerry Siegel   
2047   Steven Bochcho                            Eric Lodal   
2048      Sam Weisman                              Jay Ward   

                 Actor1              Actor2            Actor3  
0            Glenn Ford          Lee Remick   Stefanie Powers  
1            Glenn Ford          Lee Remick   Stefanie Powers  
2              Wood Moy        Marc Hayashi       Lauren Chew  
3           Roger Moore  Christopher Walken     Tanya Roberts  
4             Noah Wy

Store the combined data frame into one CSV file.

In [14]:
df.to_csv('san_francisco_film_locations.csv', index=False)

### Create an SQL Database from the Pandas Data Frame

#### Option 1: Using Duck DB
You can create a duck database using a CSV file and the **CREATE OR REPLACE TABLE** and **AS FROM** directives with the _read_csv_auto()_ method.

In [ ]:
%pip install jupysql --upgrade duckdb-engine --quiet

In [ ]:
%reload_ext sql

In [ ]:
%sql duckdb:///san_francisco_film_locations.duck.db

In [ ]:
%%sql
CREATE OR REPLACE TABLE san_francisco_film_locations AS
FROM read_csv_auto('san_francisco_film_locations.csv', header=True, sep=',')

Proceed to the section: Exploring the Database

#### Option 2: Using SQL Alchemy

In [15]:
import sqlite3 as sq3

In [16]:
conn = sq3.connect('san_francisco_film_locations.db')
#df.to_sql('san_francisco_film_locations', conn, if_exists='append', index=False)
df.to_sql('san_francisco_film_locations', conn, if_exists='replace', index=False)

In [17]:
!pip install sqlalchemy

In [18]:
%reload_ext sql

In [19]:
%sql sqlite:///san_francisco_film_locations.db

## Exploring the Database
Now that we have a database. We can start exploring it through the _SELECT_ command. _FROM_ specifies the database to query. And the '*' specifies all its contents.

A value of 5 passed to the _LIMIT_ command limits the print out to only 5 records.

In [ ]:
%%sql
SELECT *
FROM san_francisco_film_locations
LIMIT 5

These are the column attribute descriptions from the **san_francisco_film_locations** table:

|Column|Description|
|---|---|
|   Title| titles of the films|
|   ReleaseYear| time of public release of the films|
|   Locations| locations of San Francisco where the films were shot|
|   FunFacts| funny facts about the filming locations|
|   ProductionCompany| companies who produced the films|
|   Distributor| companies who distributed the films|
|   Director| people who directed the films|
|   Writer| people who wrote the films|
|   Actor1| person 1 who acted in the films|
|   Actor2| person 2 who acted in the films|
|   Actor3| person 3 who acted in the films|






## Using INSERT Statement
In this exercise, you will first go through some examples of using INSERT in queries and then solve some exercise problems by using it.


### Example 1
In this example, suppose we want to insert a new single row into the Instructor table.

1. Problem:
Insert a new instructor record with id 4 for Sandip Saha who lives in Edmonton, CA into the “Instructor” table.

2. Solution:
The following inserts the new record to the Instructor database.

In [20]:
%%sql
INSERT INTO san_francisco_film_locations(Title, ReleaseYear, Locations, FunFacts, ProductionCompany, Distributor, Director, Writer, Actor1, Actor2, Actor3)
VALUES('Some Fictitious Movie', 2024, 'Some Street in San Francisco', 'None', 'Some Production Inc.', 'Some Distributor', 'Mr. Director', 'Mr. Writer', 'Mrs. Actor', 'Mr. Actor', 'Ms. Actor');

 * sqlite:///san_francisco_film_locations.db
1 rows affected.


[]

3. Check if the new entry was included.

In [21]:
%%sql
SELECT * FROM san_francisco_film_locations;

 * sqlite:///san_francisco_film_locations.db
Done.


[('Experiment in Terror', 1962, "The Sea Captain's Chest (Fisherman's Wharf)", None, 'Columbia Pictures Corporation', 'Columbia Pictures', 'Blake Edwards', 'The Gordons', 'Glenn Ford', 'Lee Remick', 'Stefanie Powers'),
 ('Experiment in Terror', 1962, '100 St. Germain Avenue', None, 'Columbia Pictures Corporation', 'Columbia Pictures', 'Blake Edwards', 'The Gordons', 'Glenn Ford', 'Lee Remick', 'Stefanie Powers'),
 ('Chan is Missing', 1982, 'Li Po (916 Grant Avenue at Washington, Chinatown)', None, 'New Yorker Films', 'New Yorker Films', 'Wayne Wang', 'Wayne Wang', 'Wood Moy', 'Marc Hayashi', 'Lauren Chew'),
 ('A View to a Kill', 1985, "Taylor and Jefferson Streets (Fisherman's Wharf)", None, 'Metro-Goldwyn Mayer', 'MGM/UA Entertainment Company', 'John Glen', 'Richard Maibaum', 'Roger Moore', 'Christopher Walken', 'Tanya Roberts'),
 ('The Californians', 2005, None, None, 'Parker Film Company', 'Fabrication Films', 'Jonathan Parker', 'Jonathan Parker & Catherine DiNapoli', 'Noah Wyle', None, None),
 ('Babies', 2010, None, None, 'Canal+', 'Focus Features', 'Thomas Balmes', 'Thomas Balmes', 'Bayar', 'Hattie', None),
 ("I's", 2011, '1 Post Street', None, 'Banshee Cinema', None, 'Chris Edgette', 'Kyle Tuck', None, None, None),
 ('When We Rise', 2017, 'Bay Bridge', None, 'Film 49 Productions', 'Amercian Broadcasting Company', 'Gus Van Sant', 'Dustin Lance Black', 'Guy Pierce', 'Mary-Louise Parker', 'Michael Kenneth Williams'),
 ('Nash Bridges', 2021, 'California Street at Davis', None, 'Village NB Productions, LLC', 'USA Nework', 'Greg Beeman', 'Carlton Cuse, Bill Chais', 'Don Johnson', 'Cheech Marin', 'Joe Dinicol'),
 ('This Is Us', 2022, 'Alamo Square Park', None, '20th Television', 'NBC', 'Mandy Moore', 'Dan Fogelman, Casey Johnson, David Windsor, Chrissy Metz', 'Milo Ventimiglia', 'Mandy Moore', 'Sterling K. Brown'),
 ('The Rock', 1996, 'Coit Tower', 'The Tower was funded by a gift bequeathed by Lillie Hitchcock Coit, a socialite who reportedly liked to chase fires. Though the tower resembles a firehose nozzle, it was not designed this way.', 'Hollywood Pictures', 'Buena Vista Pictures', 'Michael Bay', 'David Weisberg', 'Sean Connery', 'Nicolas Cage', None),
 ('Blue Jasmine', 2013, 'Pacific & Divisadero', None, 'Perdido Productions', 'Sony Pictures Classics', 'Woody Allen', 'Woody Allen', 'Cate Blanchett', 'Alec Baldwin', 'Peter Sarsgaard'),
 ('The Caine Mutiny', 1954, 'Golden Gate Bridge', "With 23 miles of ladders and 300,000 rivets in each tower, the Golden Gate Bridge was the world's longest span when it opened in 1937.", 'Stanley Kramer Productions', 'Columbia Pictures', 'Edward Dmytryk', 'Stanley Roberts', 'Humphrey Bogart', 'Fred MacMurray', 'Jose Ferrer'),
 ('The Diary of a Teenage Girl', 2015, '700 Kansas St.', None, 'Diary the Movie, LLC', 'Sony Pictures Classics', 'Marielle Heller', 'Marielle Heller', 'Alexander Skarsgard', 'Kristen Wiig', 'Christopher Meloni'),
 ('Venom', 2018, 'Grant Ave btwn Clay and Jackson', None, 'L.O.Z. Productions, Inc.', 'Columbia Pictures, Sony Pictures Releasing', 'Ruben Fleischer', 'Jeff Pinkner, Scott Rosenberg', 'Tom Hardy', 'Michelle Wiliams', 'Riz Ahmed'),
 ('Murder in the First, Season 1', 2014, '13 Lucky Street', None, 'Turner North Center Productions', 'Turner Network Television (TNT)', 'Steven Bochcho', 'Eric Lodal', 'Taye Diggs', 'Kathleen Robertson', 'Ian Anthony Dale'),
 ('GirlBoss', 2017, 'The Castro Theater, 429 Castro St.', None, 'Hippolyta Productions, LLC', 'Netflix', 'Jamie Babbit, Amanda Brotchie, Steven K. Tsuchida, Christian Ditter, John Riggi', 'Kay Cannon', 'Britt Robertson', 'Ellie Reed', 'Amanda Rea'),
 ('Sense8 - Season 2', 2016, 'Clay between Stockton and Grant', None, 'Unpronounceable Productions, LLC', 'Netflix', 'Wachowski Siblings', 'J. Michael Straczynski, Wachowiski Siblings', 'Jamie Clayton', 'Daryl Hannah', 'Naveen Andrews'),
 ('The Pursuit of Happyness', 2006, 'Glen Park Subway Station', None, 'Columbia Pictures Corporation', 'Columbia Pictures', 'Steven C

### Example 2
In this example, suppose we want to insert some new multiple rows into the Instructor table.
1. Problem:
Insert two new instructor records into the “Instructor” table. First record with id 5 for John Doe who lives in Sydney, AU.
Second record with id 6 for Jane Doe who lives in Dhaka, BD.
2. Solution:
%%sql
INSERT INTO Instructor(ins_id, lastname, firstname, city, country)
VALUES(5, 'Doe', 'John', 'Sydney', 'AU'), (6, 'Doe', 'Jane', 'Dhaka', 'BD');
3. Copy the solution code above by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox
of the Datasette tool. Then click Submit query.
4. Copy the code below by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox of the
Datasette tool. Then click Submit query.
SELECT * FROM Instructor;

##Using UPDATE Statement
Let us go through some examples of UPDATE related queries:

### Example 3
In this example, we want to update one column of an existing row of the table.
1. Problem:
Update the city for Sandip to Toronto.
2. Solution:
UPDATE Instructor
SET city='Toronto'
WHERE firstname="Sandip";
3. Copy the solution code above by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox
of the Datasette tool. Then click Submit query.
4. Copy the code below by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox of the
Datasette tool. Then click Submit query.
SELECT * FROM Instructor;

### Example 4
In this example, we want to update multiple columns of an existing row of the table.
1. Problem:
Update the city and country for Doe with id 5 to Dubai and AE respectively.
2. Solution:
UPDATE Instructor
SET city='Dubai', country='AE'
WHERE ins_id=5;
3. Copy the solution code above by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox
of the Datasette tool. Then click Submit query.
4. Copy the code below by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox of the
Datasette tool. Then click Submit query.
SELECT * FROM Instructor;

## Using DELETE Statement
In this exercise, you will first go through an example of using DELETE in a query and then solve an exercise problem by using it.

Let us go through an example of a DELETE related query:

### Example 5
In this example, we want to remove a row from the table.
1. Problem:
Remove the instructor record of Doe whose id is 6.
2. Solution:
DELETE FROM instructor
WHERE ins_id = 6;
3. Copy the solution code above by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox
of Custom SQL query of the Datasette tool. Then click Submit query.
4. Copy the code below by clicking on the little copy button on the bottom right of the codeblock below and paste it to the textbox of the
Datasette tool. Then click Submit query.
SELECT * FROM Instructor;

## Practice Exercises

### INSERT
Now, let us practice creating and running some INSERT related queries.

#### Problem 1
Insert a new instructor record with id 7 for Antonio Cangiano who lives in Vancouver, CA into the “Instructor” table.

<details><summary>Hint</summary>
%%sql
INSERT INTO table_name([columnn1],[column2], [column3], [column4], [column5])
VALUES(7, 'Cangiano', 'Antonio', 'Vancouver', 'CA');
SELECT * FROM tablename;

```
Follow example 1 of the INSERT exercise.

```

</details>

<details><summary>Query Solution</summary>

```
 %%sql
INSERT INTO Instructor(ins_id, lastname, firstname, city, country)
VALUES(7, 'Cangiano', 'Antonio', 'Vancouver', 'CA');
SELECT * FROM Instructor;
```

</details>

#### Problem 2
Insert two new instructor records into the “Instructor” table. First record with id 8 for Steve Ryan who lives in Barlby, GB. Second
record with id 9 for Ramesh Sannareddy who lives in Hyderabad, IN.

<details><summary>Hint</summary>

```
Follow example 2 of the INSERT exercise.
%%sql
INSERT INTO table_name([columnn1],[column2], [column3], [column4], [column5])
VALUES(8, 'Ryan', 'Steve', 'Barlby', 'GB'), (9, 'Sannareddy', 'Ramesh', 'Hyderabad', 'IN');
SELECT * FROM tablename;

```

</details>

<details><summary>Query Solution</summary>

```
%%sql
INSERT INTO Instructor(ins_id, lastname, firstname, city, country)
VALUES(8, 'Ryan', 'Steve', 'Barlby', 'GB'), (9, 'Sannareddy', 'Ramesh', 'Hyderabad', 'IN');
SELECT * FROM Instructor;
```

</details>

### UPDATE
Now, let us practice creating and running some UPDATE related queries.

#### Problem 3
Update the city of the instructor record to Markham whose id is 1.

<details><summary>Hint</summary>

```
Follow example 1 of the UPDATE exercise.
%%sql
UPDATE table_name
SET [column1]='Markham'
WHERE [specifiedcolumn]=1;
SELECT * FROM tablename;
```

</details>

<details><summary>Query Solution</summary>

```
%%sql
UPDATE Instructor
SET city='Markham'
WHERE ins_id=1;
SELECT * FROM Instructor;

```

</details>

#### Problem 4
Update the city and country for Sandip with id 4 to Dhaka and BD respectively.

<details><summary>Hint</summary>

```
Follow example 2 of the UPDATE exercise.
UPDATE table_name
SET [column1]='Dhaka', [column2]='BD'
WHERE [specifiedcolumn]=4;
SELECT * FROM tablename;
```

</details>

<details><summary>Query Solution</summary>

```
%%sql
UPDATE Instructor
SET city='Dhaka', country='BD'
WHERE ins_id=4;
SELECT * FROM Instructor;
```

</details>

### DELETE
Now, let us practice creating and running a DELETE related query.

#### Problem 5
Remove the instructor record of Hima.

<details><summary>Hint</summary>

```
Follow example 1 of the DELETE exercise.
DELETE FROM tablename
WHERE [columnanme1] = 'Hima';
SELECT * FROM tablename;
```

</details>

<details><summary>Query Solution</summary>

```
%%sql
DELETE FROM instructor
WHERE firstname = 'Hima';
SELECT * FROM Instructor;
```

</details>

2. Retrieve the first 10 film names released in 2015.

<details><summary>Hint</summary>

```
Follow example 1 of LIMIT. Use DISTINCT. Use WHERE clause comparison operator =, which means Equal to.
```

</details>

<details><summary>Query Solution</summary>

```
%%sql
SELECT DISTINCT Title FROM san_francisco_film_locations WHERE ReleaseYear=2015 LIMIT 10
```

</details>

3. Retrieve the next 3 film names that follow after the first 5 films released in 2015.

<details><summary>Hint</summary>

```
Follow example 2 of the LIMIT exercise to learn how to use OFFSET. Use DISTINCT and use the WHERE clause comparison operator =, which
means Equal to.
```

</details>

<details><summary>Query Solution</summary>

```
%%sql
SELECT DISTINCT Title FROM san_francisco_film_locations WHERE ReleaseYear=2015 LIMIT 3 OFFSET 5
```

</details>

## Conclusion
Thank you for completing the lab! You are now able to perform operations on tables like:

*   Inserting rows using the INSERT command
*   Removing rows using DELETE
*   Updating the data in existing rows with UPDATE


## Credit to the Source Content
This python notebook uses the content of **Hands-on Lab : INSERT, UPDATE, DELETE** by IBM Skills Network from the [Databases and SQL for Data Science with Python](https://www.coursera.org/learn/sql-data-science) course. The source content uses a different platform, [Datasette](https://github.com/simonw/datasette), which does not utilize Jupyter notebooks.


### Change Log
All versions prior to 2.0, are attributable to IBM Skills Network's version of the **Hands-on Lab : INSERT, UPDATE, DELETE** written for [Datasette](https://github.com/simonw/datasette).


| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
|2024-05-xy|2.0|Martin Borja|Ported to Jupyter/Python Notebooks|
|2023-07-11| 1.6| Lakshmi Holla| Updated labs|
|2023-05-11| 1.6| Eric Hao & Vladislav Boyko| Updated Page Frames|
|2023-05-10| 1.5| Eric Hao & Vladislav Boyko| Updated Page Frames|
|2023-05-05| 1.4| Benny Li| Republished|
|2022-08-03| 1.3| Sathya Priya| updated HTML tag|
|2022-07-27| 1.2| Lakshmi Holla| updated HTML tag|
|2020-12-23| 1.1| Steve Ryan| ID Review|
|2020-11-30| 1.0| Sandip Saha Joy| Initial version created|

<hr>

### <h4 align="center"> **Hands-on Lab : INSERT, UPDATE, DELETE** © IBM Corporation 2023. All rights reserved. <h4/>
### <h4 align="center"> **INSERT, UPDATE, DELETE in SQL using Jupyter Notebooks** © Martin John Hilario Borja 2024. All rights reserved. <h4/>

<p>